pgmax/groups/logical.py

```diff
@@ -0,0 +1,169 @@
+"""Defines LogicalFactorGroup and its two children, ORFactorGroup and ANDFactorGroup."""
+
+import collections
+from dataclasses import dataclass, field
+from typing import FrozenSet, OrderedDict, Type
+
+import numpy as np
+
+from pgmax.factors import logical
+from pgmax.fg import groups
+
+
+@dataclass(frozen=True, eq=False)
+class LogicalFactorGroup(groups.FactorGroup):
+    """Class to represent a group of LogicalFactors.
+
+    All factors in the group are assumed to have the same edge_states_offset.
+    Consequently, the factors are all ORFactors or ANDFactors.
+
+    Args:
+        edge_states_offset: Offset to go from a variable's relevant state to its other state
+            For ORFactors the edge_states_offset is 1, for ANDFactors the edge_states_offset is -1.
+    """
+
+    log_potentials: np.ndarray = field(init=False, default=np.empty((0,)))
+    edge_states_offset: int = field(init=False)
+    factor_type: Type = field(init=False, default=logical.LogicalFactor)
+
+    def __post_init__(self):
+        # TODO: move all asserts from LogicalFactor to here
+        super().__post_init__()
+        pass
+
+    def compile_wiring(self, vars_to_starts) -> logical.LogicalWiring:
+        """Compile LogicalWiring for the LogicalFactorGroup
+
+        Args:
+            vars_to_starts: A dictionary that maps variables to their global starting indices
+                For an n-state variable, a global start index of m means the global indices
+                of its n variable states are m, m + 1, ..., m + n - 1
+
+        Returns:
+             LogicalWiring for the LogicalFactorGroup
+        """
+        # TODO: run in parallel?
+        import time
+
+        start = time.time()
+        relevant_state = (-self.edge_states_offset + 1) // 2
+
+        var_states_for_edges = []
+        for variable_and_num_states in self.variables_and_num_states:
+            variable, num_states = variable_and_num_states
+            this_var_states_for_edges = np.arange(
+                vars_to_starts[variable], vars_to_starts[variable] + num_states
+            )
+            var_states_for_edges.append(this_var_states_for_edges)
+        print(time.time() - start)
+
+        edges_num_states_cumsum = 0
+        parents_edge_states = []
+        children_edge_states = []
+        for factor_idx, variable_names_for_factor in enumerate(
```

Comment:
This seems really the same as the previous concatenation approach. The only difference is whether we create the individual factor objects. Curious to know what the performance difference is....

Also, since we are just manipulating arrays, it seems easy to make this whole function parallel using numba. Can be done in the next PR!

Reply:
This is much faster as we do not instantiate the factors objects.
Totally agree, the next PR will speed this up using numba!